File: backend/services/artist_service.go
package services

import (
	"groupie_tracker/backend/models"
	"groupie_tracker/db"
)

type ArtistService struct {
	store db.Store
}

func NewArtistService(store db.Store) *ArtistService {
	return &ArtistService{store: store}
}

func (s *ArtistService) Create(a models.Artist) models.Artist {
	return s.store.Create(a)
}

func (s *ArtistService) GetAll() []models.Artist {
	return s.store.GetAll()
}

func (s *ArtistService) GetByID(id int) (models.Artist, bool) {
	return s.store.GetByID(id)
}

func (s *ArtistService) Update(id int, a models.Artist) (models.Artist, bool) {
	return s.store.Update(id, a)
}

func (s *ArtistService) Delete(id int) bool {
	return s.store.Delete(id)
}

func (s *ArtistService) Filter(req models.FilterRequest) []models.Artist {
	return s.store.Filter(req)
}

File: frontend/src/app/services/artist.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Artist, FilterRequest } from '../models/artist.model';

@Injectable({ providedIn: 'root' })
export class ArtistService {
  private base = '/api/artists';

  constructor(private http: HttpClient) {}

  getAll(): Observable<Artist[]> {
    return this.http.get<Artist[]>(this.base);
  }

  // Query backend for initial load readiness
  getLoaded(): Observable<{ loaded: boolean }> {
    return this.http.get<{ loaded: boolean }>('/api/loaded');
  }

  getById(id: number): Observable<Artist> {
    return this.http.get<Artist>(`${this.base}/${id}`);
  }

  create(artist: Partial<Artist>) {
    return this.http.post<Artist>(this.base, artist);
  }

  update(id: number, artist: Partial<Artist>) {
    return this.http.put<Artist>(`${this.base}/${id}`, artist);
  }

  delete(id: number) {
    return this.http.delete(`${this.base}/${id}`);
  }

  search(query: string) {
    return this.http.post<Artist[]>('/api/search', { query });
  }

  filter(req: FilterRequest): Observable<Artist[]> {
    return this.http.post<Artist[]>(`${this.base}/filter`, req);
  }
}
